### ISSUE TITLE
[BUG]Zwo warmup not fully elapsed in terms of time in running workouts

### ISSUE THREAD

**Issue by lorenzogavanna** [cagnulein/qdomyos-zwift#2847]
**Describe the bug**
The warmup tag of the zwo file won't perform for the entire set duration, but it will go to next once the treadmill reaches the powerhigh value (speed).

**To Reproduce**
1. Open a zwo workout file (find it in code block later)
2. Run it
3. the speed will change but qz will not consider the time, but the power as limit in order to go to the next step of the workout.

**Expected behavior**
The duration should be the priority and not the power (speed in this case since is a running workout)-

here there is the workout i took from an old closed issue where everything seemd to work properly, so I was sure that it was not my fault
`<workout_file>
<author>Myself</author>
<name>Test Workout</name>
<description></description>
<sportType>run</sportType>
<durationType>time</durationType>
<tags></tags>
<workout>
<Warmup Duration="60" PowerLow="0.5" PowerHigh="1" pace="0"/>
<SteadyState Duration="30" Power="0.5" pace="0"/>
<SteadyState Duration="30" Power="0.5" pace="1"/>
<SteadyState Duration="30" Power="0.5" pace="2"/>
<SteadyState Duration="30" Power="0.5" pace="3"/>
<SteadyState Duration="30" Power="0.5" pace="4"/>
<Cooldown Duration="60" PowerLow="1" PowerHigh="0.5" pace="4"/>
</workout>
</workout_file>`

time to next ssems to count correctly but (in this case) when it was around 15 seconds before the end qz started the next row section of the workout.
I'm pretty sure it happens with cooldown and ramp tag as well.

Smartphone (please complete the following information):
OS: G7 thinkQ - Android 10
Mobvoi Home Treadmill Plus
QZ app 2.18.7
debug log attached.
Thank you and let me know if you need more test

[debug-Sat_Nov_30_07_55_23_2024.log](https://github.com/user-attachments/files/17964575/debug-Sat_Nov_30_07_55_23_2024.log)

**Comment by cagnulein** [comment]
Thanks I will check it ASAP

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 30 nov 2024 alle 08:09 lorenzogavanna <
***@***.***> ha scritto:

**Comment by lorenzogavanna** [comment]
I'm not at home now, but I tried with the fake treadmill setting and the issue is still there. I don't know if it's the "fake treadmill" fault or is not solved. Let me know.

**Comment by cagnulein** [comment]
I just tested and the 60 seconds warm up works fine. Maybe I didn't get the
issue?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 30 nov 2024 alle 16:17 lorenzogavanna <
***@***.***> ha scritto:

**Comment by lorenzogavanna** [comment]
Was the warmup from the code I gave you? Was your warmup performing the enitre 60 seconds and actually did the entire section in terms of power and time?  I tell you my last test:
<Warmup Duration="60" PowerLow="0.5" PowerHigh="1" pace="0"/> 
that with my settings in pace 0 (1 mile pace) means 4.2 km/h to 8.3 km/h 
it changes the speed every second (1,5 seconds) and then about 10 seconds before the end of the 60 seconds it goes to the next section. 
I tried with different pace and even without pace tag.

**Comment by cagnulein** [comment]
Ok I guess I was trying way faster, that's why I didn't see it.
Let me try again with your settings


Il giorno sab 30 nov 2024 alle 16:36 lorenzogavanna <
***@***.***> ha scritto:

**Comment by lorenzogavanna** [comment]
I saw your code changes **this is not the issue,** but you can think about implementing the Cooldown tag as well
at line 151
`if (!qstricmp(tag, "Ramp") || !qstricmp(tag, "Warmup") || !qstricmp(tag, "Cooldown")) {`
it works excactly in the same way!

**Comment by cagnulein** [comment]
fixed https://github.com/cagnulein/qdomyos-zwift/actions/runs/12114724584



> I saw your code changes **this is not the issue,** but you can think about implementing the Cooldown tag as well at line 151 `if (!qstricmp(tag, "Ramp") || !qstricmp(tag, "Warmup") || !qstricmp(tag, "Cooldown")) {` it works excactly in the same way!

no that line is fine, it's the else the cooldown case

**Comment by lorenzogavanna** [comment]
Perfectly working!! Thank you Roberto, as always!!! Let's close this as well.

**Comment by cagnulein** [comment]
amazing!
